File: mobile_tanwords/src/db/translations.ts
/**
 * Translation history — port of desktop `app/core/src/db/translations.rs`
 * plus the two translation commands that live in `settings.rs`
 * (`db_get_translation_count`, `db_clear_translations`).
 */
import { getDb } from "./connection";
import type { TranslationItem } from "@/hooks/useDB.types";

export async function db_save_translation(args: {
  sourceText: string;
  resultText: string;
  sourceLang?: string | null;
  targetLang: string;
  provider: string;
  mode: string;
}): Promise<number> {
  const db = getDb();
  const inserted = await db.runAsync(
    "INSERT INTO translations (source_text, result_text, source_lang, target_lang, provider, mode) VALUES (?, ?, ?, ?, ?, ?)",
    [args.sourceText, args.resultText, args.sourceLang ?? null, args.targetLang, args.provider, args.mode]
  );

  await db.runAsync(
    "INSERT INTO daily_streaks (date, translations) VALUES (date('now'), 1) " +
      "ON CONFLICT(date) DO UPDATE SET translations = translations + 1"
  );

  return inserted.lastInsertRowId;
}

export async function db_get_translations(args?: {
  search?: string | null;
  cluster?: string | null;
}): Promise<TranslationItem[]> {
  const db = getDb();

  let sql =
    "SELECT id, source_text, result_text, source_lang, target_lang, provider, mode, cluster_tag, created_at " +
    "FROM translations WHERE 1=1";

  const values: string[] = [];

  const search = args?.search ?? null;
  if (search !== null) {
    // One shared bind in Rust (?N twice) → two binds here.
    sql += " AND (source_text LIKE ? OR result_text LIKE ?)";
    values.push(`%${search}%`, `%${search}%`);
  }

  const cluster = args?.cluster ?? null;
  if (cluster !== null) {
    sql += " AND cluster_tag = ?";
    values.push(cluster);
  }

  sql += " ORDER BY created_at DESC LIMIT 200";

  return db.getAllAsync<TranslationItem>(sql, values);
}

export async function db_get_translation_count(): Promise<number> {
  const db = getDb();
  const row = await db.getFirstAsync<{ n: number }>("SELECT COUNT(*) AS n FROM translations");
  return row?.n ?? 0;
}

export async function db_clear_translations(): Promise<void> {
  const db = getDb();
  await db.runAsync("DELETE FROM translations");
}

File: app/core/src/db/translations.rs
use crate::db::params; use crate::db::Value;
use serde::Serialize;
use crate::shim::State;

use crate::AppState;
use crate::db;

#[derive(Serialize)]
pub struct TranslationItem {
    pub id: i64,
    pub source_text: String,
    pub result_text: String,
    pub source_lang: String,
    pub target_lang: String,
    pub provider: String,
    pub mode: String,
    pub cluster_tag: Option<String>,
    pub created_at: String,
}

#[crate::shim::command]
pub async fn db_save_translation(
    source_text: String,
    result_text: String,
    source_lang: Option<String>,
    target_lang: String,
    provider: String,
    mode: String,
    conn: State<'_, AppState>,
) -> Result<i64, String> {
    let db = db::conn(&conn)?;
    let id = db::fetch_one(
        &db,
        "INSERT INTO translations (source_text, result_text, source_lang, target_lang, provider, mode) VALUES (?1, ?2, ?3, ?4, ?5, ?6) RETURNING id",
        params![source_text, result_text, source_lang, target_lang, provider, mode],
        |r| r.get::<i64>(0),
    )
    .await?;

    db.execute(
        "INSERT INTO daily_streaks (\"date\", translations) VALUES (date('now'), 1)
         ON CONFLICT(\"date\") DO UPDATE SET translations = translations + 1",
        (),
    )
    .await
    .map_err(|e| e.to_string())?;

    Ok(id)
}

#[crate::shim::command]
pub async fn db_get_translations(
    search: Option<String>,
    cluster: Option<String>,
    conn: State<'_, AppState>,
) -> Result<Vec<TranslationItem>, String> {
    let db = db::conn(&conn)?;

    let mut sql = String::from(
        "SELECT id, source_text, result_text, source_lang, target_lang, provider, mode, cluster_tag, created_at
         FROM translations WHERE 1=1"
    );

    let mut param_values: Vec<Value> = vec![];

    if let Some(ref s) = search {
        let idx = param_values.len() + 1;
        sql.push_str(&format!(" AND (source_text LIKE ?{idx} OR result_text LIKE ?{idx})"));
        param_values.push(Value::from(format!("%{}%", s)));
    }

    if let Some(ref c) = cluster {
        let idx = param_values.len() + 1;
        sql.push_str(&format!(" AND cluster_tag = ?{idx}"));
        param_values.push(Value::from(c.clone()));
    }

    sql.push_str(" ORDER BY created_at DESC LIMIT 200");

    db::fetch_all(&db, &sql, param_values, |row| {
        Ok(TranslationItem {
            id: row.get(0)?,
            source_text: row.get(1)?,
            result_text: row.get(2)?,
            source_lang: row.get(3)?,
            target_lang: row.get(4)?,
            provider: row.get(5)?,
            mode: row.get(6)?,
            cluster_tag: row.get(7)?,
            created_at: row.get(8)?,
        })
    })
    .await
}
